src/charm.py

```diff
@@ -239,6 +253,83 @@ def _get_hydra_endpoint_info(self) -> Optional[str]:
+        if identity_schemas := self.config.get("identity_schemas"):
+            default_schema = self.config.get("default_identity_schema")
+            if not default_schema:
+                logger.error(
+                    "`identity_schemas` configuration was set, but no `default_identity_schema` was found"
+                )
+                logger.warning("Ignoring `identity_schemas` configuration")
+                return None
+
+            try:
+                schemas = json.loads(identity_schemas)
+            except json.JSONDecodeError as e:
+                logger.error(f"identity_schemas is not a valid json: {e}")
+                logger.warning("Ignoring `identity_schemas` configuration")
+                return None
+
+            schemas = {
+                schema_id: f"base64://{base64.b64encode(json.dumps(schema).encode()).decode()}"
+                for schema_id, schema in schemas.items()
+            }
+            return default_schema, schemas
+        return None
+
+    def _get_shared_identity_schema_config(self) -> Optional[Tuple[str, Dict]]:
+        storage = self.model.storages[SHARED_IDENTITY_SCHEMA_STORAGE_NAME]
+        if not storage:
+            logger.info("No shared storage attached")
+            return
+
+        path = storage[0].location
+        schemas = {
+            schema_file.stem: f"file://{self._identity_schemas_shared_dir_path / schema_file.name}"
+            for schema_file in path.glob("*.json")
+        }
+        if not schemas:
+            return None
+        default_schema_id_file = (
+            self._identity_schemas_local_dir_path / DEFAULT_SCHEMA_ID_FILE_NAME
+        )
+        with open(default_schema_id_file) as f:
+            default_schema = f.read()
+        if default_schema not in schemas:
+            return None
+        return default_schema, schemas
+
+    def _get_default_identity_schema_config(self) -> Tuple[Optional[str], Optional[Dict]]:
+        schemas = {
+            schema_file.stem: f"file://{self._identity_schemas_default_dir_path / schema_file.name}"
+            for schema_file in self._identity_schemas_local_dir_path.glob("*.json")
+        }
+        default_schema_id_file = (
+            self._identity_schemas_local_dir_path / DEFAULT_SCHEMA_ID_FILE_NAME
+        )
+        with open(default_schema_id_file) as f:
+            default_schema = f.read()
+        if default_schema not in schemas:
+            raise RuntimeError(f"Default schema `{default_schema}` can't be found")
+        return default_schema, schemas
+
+    def _get_identity_schema_config(self) -> Optional[Tuple[str, Dict]]:
+        """Get the the default schema id and the identity schemas.
+
+        The identity schemas can come from various sources. We chose them in this order:
+        1) If the user has defined some schemas in the juju config, return those
+        2) Else if there are identity schemas in the shared folder that were
```

Comment:
Is there a reason/usecase why we want this option 2?

Reply:
Kratos offers no API for updating the identity schemas, this is how we decided to overcome this for the time being:
- The admin login UI will place the schemas provided from the user in the shared volume
- The kratos charm will scan the shared folder and update the kratos config if any schema is found in the shared volume.

Reply:
The shared volume is shared by the login UI and Kratos operators? If that's the case why not use relations?

Reply:
The admin UI (which does not exist yet) and the Kratos container. We can't use relations because we want to the user to define some schema in the admin UI and then push that schema in the kratos container either in the form of a file or base64 encoded in the the kratos config.

If we were to use relations for this we would have to have a flow like the following:
- User defines schema in the admin UI and store it somewhere.
- The admin UI charm calls some API of the admin UI to fetch the schemas and updates the relation databag.
- The kratos charm listens for changes in the relation databag and updates the kratos config.

This flow seems more complicated and we have to keep 3 different storages in sync (the admin UI storage, the relation databag and the kratos storage).

Also both solutions suffer from the same problem that we can't trigger the kratos operator to update its config once a schema is altered in the admin UI. (Perhaps we could use a pebble check for this?)

We know that this approach is not optimal and problems may arise on the way but implementing the admin ui is far in the future so we can re-iterate on this when the time comes. Any suggestions are more than welcome.

Reply:
All right, I'll resolve the comment, is this a spec somewhere?

Reply:
There is a brief description on https://docs.google.com/document/d/13LKn-LmFom1ahUNzVFz7eU3G2aEQN4FpwuZxhB8QM8Q/edit#heading=h.q48gjgxr36ad. There is not much detail there yet.

Reply:
I guess I'm still unsure as to why we implemented this here if we're not sure how/when this will be set up. If we know we won't need it right now and there's no way to test it, I'd wait it out.

Reply:
I have no problem with that, I will remove it for now.